perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java

```diff
@@ -197,37 +197,46 @@ public static boolean isAuthorizedForAttribute(PerunSession sess, ActionType act
 
 		//Important: There is no options for other roles like service, serviceUser and other!
 		if (resource != null && member != null) {
-			if (roles.contains(Role.VOADMIN)) {
+			if (roles.containsKey(Role.VOADMIN)) {
 				if (isAuthorized(sess, Role.VOADMIN, member)) return true;
 			}
-			if (roles.contains(Role.VOOBSERVER)) {
+			if (roles.containsKey(Role.VOOBSERVER)) {
 				if (isAuthorized(sess, Role.VOOBSERVER, member)) return true;
 			}
-			if (roles.contains(Role.FACILITYADMIN)) {
+			if (roles.containsKey(Role.FACILITYADMIN)) {
 				if (isAuthorized(sess, Role.FACILITYADMIN, resource)) return true;
 			}
-			if (roles.contains(Role.SELF)) {
-				if (isAuthorized(sess, Role.SELF, member)) return true;
+			if (roles.containsKey(Role.SELF)) {
+				if (roles.get(Role.SELF).contains(ActionType.READ_PUBLIC) || roles.get(Role.SELF).contains(ActionType.WRITE_PUBLIC)) return true;
+				if (roles.get(Role.SELF).contains(ActionType.READ) || roles.get(Role.SELF).contains(ActionType.WRITE)) {
+					if (isAuthorized(sess, Role.SELF, user)) return true;
+				}
+				if (roles.get(Role.SELF).contains(ActionType.READ_VO) || roles.get(Role.SELF).contains(ActionType.WRITE_VO)) {
+					Vo v = getPerunBl().getMembersManagerBl().getMemberVo(sess, member);
+					if (getPerunBl().getResourcesManagerBl().getResources(sess, v).contains(resource)) {
```

Comment:
I believe this check is not sufficient. You get here member and resource from the attribute you are trying to access. Then you verifies, that the Vo of this member corresponds to one of Vos that can access the resource. That's first part of the check. ( You also do not need to get the actual Vo object, since the member contains its Vo's id.)

But you also need to verify that the the current principal is a member of the same Vo as the member from the attribute.